### ISSUE TITLE
Required format for printing external loads files can't be met (Matlab)

### ISSUE THREAD

**Issue by jimmyDunne** [opensim-org/opensim-core#1650]
When printing out forces from a c3d, you have to rely on `STOFileAdapter()`, which prints external loads in a [non-compliant .mot format](http://simtk-confluence.stanford.edu:8080/display/OpenSim/Motion+%28.mot%29+Files). 

```
%% import java libraries
import org.opensim.modeling.*

%% Use a c3dAdapter to turn read a c3d file
adapter = C3DFileAdapter();
forces = adapter.read(filepath).tables.get('forces');

%% flatten the Vec3 table to a table of doubles. 
postfix = StdVectorString(); postfix.add('_x');postfix.add('_y');postfix.add('_z');
forces_flattened = forces_rot.flatten(postfix);
% make a sto adapter and write the forces table to file.
STOFileAdapter().write(forces_flattened,[filepath '.mot']);
```

The output file header looks like this

```
CalibrationMatrices=Corners=DataRate=2000.000000
Origins=Types=events=DataType=double
version=2
endheader
time	f1_x	f1_y	f1_z	m1_x	m1_y	m1_z	p1_x	p1_y	p1_z	f2_x	f2_y	f2_z	m2_x	m2_y	m2_z	p2_x	p2_y	p2_z
0	1.472130060195923	17.71418762207031	-2.357758045196534	61.67040961382952	742.0819913723619	-98.77109934038099	827.2036135380047	3.6909246695582e-14	602.7737421316849	0	22.82806015014648	0.04444999992847303	2.273736754432321e-13	814.478227169007	1.585923503849422	318.8885687585697	3.419024461569678e-14	558.3690683632433
```

There are a few issues with this. (i) the header information is not compliant with standard .mot file format, and (ii) the order of colomns is also incorrect (we require f1,p1,f2,p2,...fn,pn, m1,m2,...mn).

I should be able to open the .mot in the (3.3) GUI to test/visualize, which I currently can not do.

**Comment by klshrinidhi** [comment]
* Something is definitely wrong with writing the header. I will try fixing that.
* About column order, it may be better if I write a method for DataTable to rearrange columns internally. The order of columns from C3DFileAdapter seems logical -- F,1 M1, P1, F2, M2, P2, ..... , Fn, Mn, Pn. So one could rearrange the columns of the forces table coming out of C3DFileAdapter using something like:
`table.reorderColumns({1, 2, 5, 4, 3})` -- reverses only the last 3 columns of the table.

**Comment by jenhicks** [comment]
Additional steps discussed at dev meeting:
1) - add flag to C3DFileAdapter to output force mot in traditional format (default to new format)
Flag Name: UseTraditionalMotFormat (open to other suggestions)
2)  add a MotFileAdapter: compliant mot header format

**Comment by tkuchida** [comment]
I suggest `useSIMMColumnOrdering` because (1) "traditional" is ambiguous, (2) the .mot format is specific to SIMM (based on [this](http://simtk-confluence.stanford.edu:8080/display/OpenSim/Motion+%28.mot%29+Files) Confluence page), and (3) IIUC the only difference will be the column ordering.

**Comment by jenhicks** [comment]
I like this suggestion @tkuchida  Is the format specific to SIMM or Motion Analysis products in general? @jimmyDunne @aymanhab ?

**Comment by aymanhab** [comment]
Yes @jenhicks AFAIK. Also keep in mind that C-Motion processes files into this format when they export mocap data.

**Comment by jenhicks** [comment]
(1) Add TimeSeriesTable to Storage converter; private member function on Storage. 
(2) Storage constructor that takes a filename uses this utility

**Comment by aymanhab** [comment]
@aseth1 will follow up with @klshrinidhi

**Comment by jenhicks** [comment]
@jimmyDunne can you supply materials for a test case

**Comment by aseth1** [comment]
1. The order of the columns is not the issue nor is a SIMMMotionFile header required. An .mot is only required because it seems to be what the GUI uses to decide that the file contains GRFs. In other words, the previewer will assume GRFs if it loads a *.mot*. Loading ant *.sto* file leads to an error (which is really a bug in 3.3).

2. What appears to be strictly required is that the columns labels use *_v(x/y/z)* for the components of the force vector and *_p(x/y/z)* for the force point coordinates. The first *p#* and *v#* triplets are paired together and any subsequent set of triplets are paired to draw a vector emerging from the point. Torques are ignored, not required and not previewed.

IMHO it is simpler to have the conversion scripts adopt the same naming scheme (_vx, _px, etc...)  than adding a SIMMMotion *exporter*. @jimmyDunne can you try that to see if it works? In 4.0 we can make that requirement go away by having the previewer use the ExternalLoads class directly.

**Comment by jenhicks** [comment]
@jimmyDunne Is working on some matlab scripts to test this. Emailed @aseth1 with a couple questions. Short version is that we can probably deal with it on the Matlab side.

**Comment by jimmyDunne** [comment]
@aseth1 STOFileAdapter writes a (forces) file with the below header;

```
CalibrationMatrices=Corners=DataRate=2000.000000
Origins=Types=events=DataType=double
version=2
endheader
```
Storage class can not read that header information. I can open the file if I (manually) change the header to this;

```
nColumns=19
nRows=641
endheader
```
The STOFileAdpater should be updated.

**Comment by chrisdembia** [comment]
@jimmyDunne are you trying to load the file with the Storage class or with the STOFileAdapter class?

**Comment by jimmyDunne** [comment]
The STOFileAdapter() does the writing, the Storage() class does the reading. The exception occurs in 4.0 but it seems that this is also an issue in previous versions.

**Comment by chrisdembia** [comment]
What if you load the file with the STOFileAdapter?

We were planning on modifying the constructor to Storage so that it could read version-2 storage files by using the STOFileAdapter. I'm not sure whose plate that's on.

**Comment by jimmyDunne** [comment]
`STOFileAdapter().read()` works. When usning Storage, I get;

```Java exception occurred:
java.io.IOException: Storage: ERROR- failed to parse headers of file test_walking.mot
	at org.opensim.modeling.opensimCommonJNI.new_Storage__SWIG_4(Native Method)
	at org.opensim.modeling.Storage.<init>(Storage.java:97)
```

**Comment by aseth1** [comment]
The `Storage` reading a version 2 file, in 4.0 is a good idea to ease the transition, but it won't help trying to read .sto and .mot files into the GUI since it appears that the GUI motion previewer has its won reader that expects:
```
nColumns=19
nRows=641
endheader
```
Perhaps we can add an `writeAsVersionZero` on `STOFileAdapter` that flattens the Table and writes out the expected header for the GUI (at least in 3.3) and others (like several old MATLAB scripts) that use the header numbers.

**Comment by aymanhab** [comment]
@aseth1 GUI does no parsing of its own, it uses the constructor of the Storage class as is.

**Comment by aseth1** [comment]
Thanks Ayman. For some reason I thought `nColumns`, `nRows` were not a strict requirement of Storage. I was wrong.  It is that Storage takes a multitude of number of columns/rows descriptors, like `nc`, `nCols` and `datacolumns`.

Nonetheless, a `writeAsVerisonZero` (or `One`) that writes out one variant of these required tags, would work. Alternatively, the MATLAB script using the `STOFileAdapter` could insert these as metadata. That could work too.

**Comment by aymanhab** [comment]
No worries, thanks @aseth1

**Comment by jimmyDunne** [comment]
I have managed to hack the column/row numbering on the Matlab side by editing (deleting all the existing values and adding nColumn and nRow entries) the meta key values. 

I can now print a header that looks like  this;
```
nColumns=19
nRows=6448
DataType=double
version=2
endheader
```

This is a hack on the Matlab side that should work for now. But there needs to be a fix for how the headers get printed in the sto files. 

I am still getting an exception in the OpenSim 3.3 GUI when trying to view the .mot file. I have attached it here (as a zip). @aymanhab would you mind having a look to see if you can preview the forces in 3.3?
[test_walking.mot.zip](https://github.com/opensim-org/opensim-core/files/1120748/test_walking.mot.zip)

**Comment by aseth1** [comment]
@jimmyDunne I'll take a look.

**Comment by aseth1** [comment]
@jimmyDunne I have the file load in 3.3 but there is nothing being shown. Should there be forces being drawn?

**Comment by jimmyDunne** [comment]
Yep.

**Comment by aseth1** [comment]
I think the issue was addressed in #1843 and was verified in #1925. @jimmyDunne that resulting *forces* files were loaded and successfully previewed into to 3.3. Thanks!
